Expand bi-directional associations guide [ci skip]

composerinteralia opened the pull request:
I remember being confused about what exactly association inverses would
do for me. This guide helped, but it leaves out some use cases. I'm hoping
that having all these use cases documented in one place with code
examples for each will make it easier for folks to understand why
they might want to set `:inverse_of` when Rails can't infer it.

Specifically, this commit:

* Adds an example explicitly demonstrating that all the `Author` objects are the same, and that no additional queries are executed.
* Preserves the example around preventing inconsistent data (although replaces the single-letter variable names)
* Adds an example around autosaving, which I'm not sure is explicitly documented elsewhere
* Adds an example around validations, linking to the relevant sections of the validation guide where we mention needing to set the `:inverse_of` for these validations

This commit also adds links from the `:inverse_of` sections back up to
the Bi-directional associations sections, hopefully making those
`:inverse_of` sections clearer.

kamipo commented on guides/source/association_basics.md:
````diff
@@ -761,15 +761,53 @@ class Book < ApplicationRecord
+
+```irb
+irb> author = Author.first
+irb> authors = author.books.all? do |book|
````
```suggestion
irb> author.books.all? do |book|
```

composerinteralia replied:
Nice catch!

jonathanhefner commented on guides/source/association_basics.md:
````diff
@@ -761,15 +761,53 @@ class Book < ApplicationRecord
+
+* Prevent needless queries for already-loaded data
+
+```irb
+irb> author = Author.first
+irb> author.books.all? do |book|
+irb>   book.author.equal?(author) # No additional queries executed here
+irb> end
+=> true
+```
````
If we want to maintain the bulleted list margin, we should indent these code snippets:

```suggestion
    ```irb
    irb> author = Author.first
    irb> author.books.all? do |book|
    irb>   book.author.equal?(author) # No additional queries executed here
    irb> end
    => true
    ```
```

jonathanhefner commented on guides/source/association_basics.md:
````diff
@@ -761,15 +761,53 @@ class Book < ApplicationRecord
+=> true
+```
+
+* Validate the [presence](active_record_validations.html#presence) and [absence](active_record_validations.html#absence) of associations in more cases
````
I don't have a strong opinion about wrapping markdown, but, for consistency with the rest of this section:

```suggestion
* Validate the [presence](active_record_validations.html#presence) and 
  [absence](active_record_validations.html#absence) of associations in more cases
```

jonathanhefner commented on guides/source/association_basics.md:
````diff
@@ -761,15 +761,53 @@ class Book < ApplicationRecord
+* Validate the [presence](active_record_validations.html#presence) and [absence](active_record_validations.html#absence) of associations in more cases
+
+```irb
+irb> author = Author.new
+irb> book = author.books.new
+irb> book.author.present?
+=> true
+irb> book.valid?
 => true
````
Perhaps a counter example here, to further illustrate the point?

```suggestion
irb> book = Book.new
irb> book.author.present?
=> false
irb> book.valid?
=> false
irb> author = Author.new
irb> book = author.books.new
irb> book.author.present?
=> true
irb> book.valid?
=> true
```

Or maybe with validation error messages instead?

```suggestion
irb> book = Book.new
irb> book.valid?
=> false
irb> book.errors.full_messages
=> ["Author must exist"]
irb> author = Author.new
irb> book = author.books.new
irb> book.valid?
=> true
```

jonathanhefner commented on guides/source/association_basics.md:
```diff
@@ -761,15 +761,53 @@ class Book < ApplicationRecord
 => true
-irb> a.first_name = 'David'
-irb> a.first_name == b.author.first_name
+irb> author.name = 'Changed Name'
```
```suggestion
irb> author.name = "Changed Name"
```

jonathanhefner commented on guides/source/association_basics.md:
```diff
@@ -793,15 +831,51 @@ class Book < ApplicationRecord
+irb> book = author.books.first
+irb> author.name == book.author.name
+=> true
+irb> author.name = 'Changed Name'
```
```suggestion
irb> author.name = "Changed Name"
```

jonathanhefner commented on guides/source/association_basics.md:
````diff
@@ -793,15 +831,51 @@ class Book < ApplicationRecord
+* Fail to validate presence or absence
+
+```irb
+irb> author = Author.new
+irb> book = author.books.new
+irb> book.author.present?
+=> false
+irb> book.valid?
 => false
````
If you do update the snippet from line 806 to show validation error messages, this snippet should be updated to match.

jonathanhefner reviewed (commented):
Nice! :+1:

composerinteralia commented:
Thank you! Updated to include these suggestions.

jonathanhefner commented on guides/source/association_basics.md:
````diff
@@ -761,17 +761,60 @@ class Book < ApplicationRecord
+
+* Prevent needless queries for already-loaded data
+
+  ```irb
+  irb> author = Author.first
+  irb> author.books.all? do |book|
+  irb>   book.author.equal?(author) # No additional queries executed here
+  irb> end
+  => true
+  ```
````
Unfortunately, markdown doesn't recognize the code block as being part of the list item if it only has a 2-space indent.  It has to be at least 4 spaces:

```suggestion
    ```irb
    irb> author = Author.first
    irb> author.books.all? do |book|
    irb>   book.author.equal?(author) # No additional queries executed here
    irb> end
    => true
    ```
  ```
```

composerinteralia replied:
🤦🏻 Thanks for spotting that!

composerinteralia committed:
Expand bi-directional associations guide [ci skip]

I remember being confused about what exactly association inverses would
do for me. This guide helped, but it leaves out some use cases. I'm hoping
that having all these use cases documented in one place with code
examples for each will make it easier for folks to understand why
they might want to set `:inverse_of` when Rails can't infer it.

Specifically, this commit:

* Adds an example explicitly demonstrating that all the `Author` objects
  are the same, and that no additional queries are executed.
* Preserves the example around preventing inconsistent data (although
  replaces the single-letter variable names)
* Adds an example around autosaving, which I'm not sure is explicitly
  documented elsewhere
* Adds an example around validations, linking to the relevant sections
  of the validation guide where we mention needing to set the
`:inverse_of` for these validations

This commit also adds links from the `:inverse_of` sections back up to
the Bi-directional associations sections, hopefully making those
`:inverse_of` sections clearer.

jonathanhefner commented:
Thank you, @composerinteralia! :wave:
